### PR TITLE
Clean up `Device.batch_transform` to use `split_non_commuting`

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**
https://github.com/PennyLaneAI/pennylane/pull/5729 introduced the unified `split_non_commuting`. Now the legacy device can use `split_non_commuting` in all scenarios.

**Description of the Change:**
Cleans up `batch_transform` to use `split_non_commuting`

**Benefits:**
Cleaner code, prepares for the deprecation of `hamiltonian_expand` and `sum_expand`.

**Related GitHub Issues:**
[sc-61253]

### CODE REVIEW

File: pennylane/_device.py [hunk 1]
```diff
@@ -738,79 +734,91 @@ def batch_transform(self, circuit: QuantumTape):
...
+        # i.e. while we first applied `split_non_commuting` to the tape, we need to process the
         # results from the broadcast expansion first.
         def total_processing(results):
-            return hamiltonian_fn(expanded_fn(results))
+            return processing_fn(expanded_fn(results))
 
         return expanded_tapes, total_processing
 
+    def _all_multi_term_obs_supported(self, circuit):
+        """Check whether all multi-term observables are supported."""
```
Comment (dwierichs) [inline pennylane/_device.py:802]:
```suggestion
        """Check whether all multi-term observables in the circuit are supported."""
```

File: pennylane/_device.py [hunk 2]
```diff
@@ -738,79 +734,91 @@ def batch_transform(self, circuit: QuantumTape):
...
+
+        for obs in circuit.observables:
+
+            if not hasattr(obs, "name"):
```
Comment (dwierichs) [inline pennylane/_device.py:806]: Could you comment in the code on when this clause is reached?

File: pennylane/_device.py [hunk 3]
```diff
@@ -738,79 +734,91 @@ def batch_transform(self, circuit: QuantumTape):
...
         finite_shots = self.shots is not None
-        grouping_known = all(
-            obs.grouping_indices is not None
-            for obs in circuit.observables
-            if isinstance(obs, (Hamiltonian, LinearCombination))
+        is_shadow = any(isinstance(m, ShadowExpvalMP) for m in circuit.measurements)
+        is_analytic_or_shadow = not finite_shots or is_shadow
+        all_obs_usable = self._all_multi_term_obs_supported(circuit)
+        exists_multi_term_obs = any(
+            isinstance(m.obs, (Hamiltonian, Sum, Prod, SProd)) for m in circuit.measurements
```
Comment (dwierichs) [inline pennylane/_device.py:746]: Should `LinearCombination` appear in this tuple? :thinking:

File: tests/devices/default_qubit/test_default_qubit_tracking.py [hunk 4]
```diff
@@ -198,7 +198,11 @@ def circuit_3(y):
     ([qml.expval(qml.PauliX(0))], 1, 10),
     ([qml.expval(qml.PauliX(0)), qml.expval(qml.PauliY(0))], 2, 20),
     # Hamiltonian test cases
-    ([qml.expval(qml.Hamiltonian([1, 1], [qml.PauliX(0), qml.PauliX(1)]))], 2, 20),
+    (
+        [qml.expval(qml.Hamiltonian([1, 0.5, 1], [qml.PauliX(0), qml.PauliY(0), qml.PauliX(1)]))],
+        2,
+        20,
+    ),
```
Comment (dwierichs) [inline tests/devices/default_qubit/test_default_qubit_tracking.py:205]: Cosmetics...

```suggestion
    ([qml.expval(qml.Hamiltonian([1, 0.5, 1], [qml.X(0), qml.Y(0), qml.X(1)]))], 2, 20),
```

File: tests/interfaces/test_autograd.py [hunk 5]
```diff
@@ -200,47 +199,21 @@ def test_no_batch_transform(self, mocker):
...
 
         spy.assert_not_called()
 
         res = qml.execute([tape], dev, None, device_batch_transform=True)
         spy.assert_called()
 
-        assert isinstance(res[0], np.ndarray)
-        assert res[0].shape == ()
-        assert np.allclose(res[0], np.cos(y), atol=0.1)
+        # assert isinstance(res[0], np.ndarray)
```
Comment (dwierichs) [inline tests/interfaces/test_autograd.py:214]: Is this supposed to go in or be deleted?

File: tests/test_device.py [hunk 6]
```diff
@@ -1191,9 +1191,9 @@ def test_batch_transform_does_not_expand_supported_sum(self, mocker):
         """Tests that batch_transform does not expand Sums if they are supported."""
         H = qml.sum(qml.PauliX(0), qml.PauliY(0))
         qs = qml.tape.QuantumScript(measurements=[qml.expval(H)])
-        spy = mocker.spy(qml.transforms, "sum_expand")
+        spy = mocker.spy(qml.transforms, "split_non_commuting")
 
-        dev = self.SomeDevice()
+        dev = self.SomeDevice(shots=None)
```
Comment (dwierichs) [inline tests/test_device.py:1196]: Why is this added?

File: pennylane/_device.py [hunk 7]
```diff
@@ -738,79 +734,91 @@ def batch_transform(self, circuit: QuantumTape):
...
+        def null_postprocess(results):
+            return results[0]
+
         finite_shots = self.shots is not None
-        grouping_known = all(
-            obs.grouping_indices is not None
-            for obs in circuit.observables
-            if isinstance(obs, (Hamiltonian, LinearCombination))
+        is_shadow = any(isinstance(m, ShadowExpvalMP) for m in circuit.measurements)
+        is_analytic_or_shadow = not finite_shots or is_shadow
```
Comment (dwierichs) [inline pennylane/_device.py:743]:
```suggestion
        has_shadow = any(isinstance(m, ShadowExpvalMP) for m in circuit.measurements)
        is_analytic_or_shadow = not finite_shots or has_shadow
```

**Review by dwierichs — CHANGES_REQUESTED** [review]
Nice @astralcai !
Took me some time to dig through the logic, but it is nicer now! :) 
Also good to see the reduction of number of terms in some scenarios! :tada: 

Had only a few quite small comments, basically ready to approve.

File: pennylane/_device.py (hunk 3 above)
Comment (astralcai) [reply pennylane/_device.py:746]: `LinearCombination` extends `Sum`

File: tests/test_device.py (hunk 6 above)
Comment (astralcai) [reply tests/test_device.py:1196]: The original test was incorrect in this case, because the legacy device would only support `Sum` in analytic mode, not with finite shots. This is the case for `Hamiltonian`, so it should be the same for `Sum` and other multi-term observables too.

File: pennylane/_device.py (hunk 2 above)
Comment (astralcai) [reply pennylane/_device.py]: I updated the code so that it's easier to read now.

File: tests/test_device.py (hunk 6 above)
Comment (dwierichs) [reply tests/test_device.py:1196]: I see! And when not providing shots, `SomeDevice` defaults to finite shots? :thinking:

File: pennylane/_device.py [hunk 8]
```diff
@@ -738,79 +734,92 @@ def batch_transform(self, circuit: QuantumTape):
...
         return expanded_tapes, total_processing
 
+    def _all_multi_term_obs_supported(self, circuit):
+        """Check whether all multi-term observables in the circuit are supported."""
+
+        for mp in circuit.measurements:
+
+            if mp.obs is None:
+                # Some measurements are not observable based.
+                continue
```
Comment (dwierichs) [inline pennylane/_device.py:808]: Ah, yeah I definitely did not get that before, this condition is nice :)

File: tests/test_device.py (hunk 6 above)
Comment (astralcai) [reply tests/test_device.py:1196]: Yes, to 1000 shots.

Comment (dwierichs) [reply tests/test_device.py:1196]: :o I was not aware of this default in the core library _at all_. Thanks!

**Review by dwierichs — APPROVED** [review]
Thanks for the clarifications! :+1: 
LGTM! :rocket:

**Review by PietropaoloFrisoni — APPROVED** [review]
LGTM!
